### PR TITLE
Fix handlers_test linkage and windows failures

### LINKED ISSUE #380: rust_test / handlers_test is broken on windows.
As described [here](https://github.com/ry/deno/pull/371#issuecomment-405607736) 
```
= note: handlers_test.handlers_test0.rcgu.o : error LNK2019: unresolved external symbol deno_reply_error referenced in function handle_code_fetch
handlers_test.handlers_test0.rcgu.o : error LNK2019: unresolved external symbol deno_reply_code_fetch referenced in function handle_code_fetch
handlers_test.exe : fatal error LNK1120: 2 unresolved externals
```

### PR DESCRIPTION
**Pull request by piscisaureus**
Tested on linux and windows, but not on mac.

Fixes #380

### CODE REVIEW

File: BUILD.gn [hunk 1]
```diff
@@ -20,6 +20,12 @@ config("deno_config") {
   if (is_debug) {
     defines = [ "DEBUG" ]
   }
+  if (is_mac) {
+    libs = [ "resolv" ]
+  }
+  if (is_win) {
+    libs = [ "userenv.lib" ]
+  }
```
Comment (ry) [inline BUILD.gn]: Are these not discovered with the new script?

File: BUILD.gn [hunk 2]
```diff
@@ -63,6 +69,21 @@ rust_test("handlers_test") {
...
+    # * Linking to `:handlers` or `:libdeno` isn't possible, because they
+    #   already contain some symbols exported by `handlers.rs`. These duplicate
+    #   symbols trip up the linker.
+    # * The `rust_test` and `rust_executable` templates only produce an object
+    #   file, and then invoke an external linker. Transitive rust depencenies
+    #   are not resolved in either step.
+    "$rust_build:idna",
+    "$rust_build:percent_encoding",
+    "$rust_build:unicode_bidi",
+    "$rust_build:unicode_normalization",
```
Comment (ry) [inline BUILD.gn]: This is unfortunate. What if we moved all the tests into their own file - `handlers_test.rs` ?

File: build_extra/rust/get_rust_ldflags.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env python
+# Copyright 2018 Bert Belder <bertbelder@gmail.com>
```
Comment (ry) [inline build_extra/rust/get_rust_ldflags.py]: We should change to "Deno authors" soon. Maybe after we move out of github.com/ry

Please add a big comment at the top of this file explaining the situation... I think this is likely to be confusing to future readers.

File: build_extra/rust/rust.gni [hunk 4]
```diff
@@ -1,4 +1,12 @@
-stdlib_label = "//build_extra/rust:stdlib"
+empty_rs_path = rebase_path("empty.rs", root_build_dir)
+rust_bin_ldflags =
+    exec_script("get_rust_ldflags.py", [ empty_rs_path ], "list lines")
+rust_test_ldflags = exec_script("get_rust_ldflags.py",
+                                [
+                                  empty_rs_path,
+                                  "--test",
+                                ],
+                                "list lines")
```
Comment (ry) [inline build_extra/rust/rust.gni]: list lines?

Please point the reader to the soon-to-be-written comment at the top of get_rust_ldflags.py

File: build_extra/rust/rust.gni [hunk 5]
```diff
@@ -47,43 +55,30 @@ template("run_rustc") {
...
+      emit_type = "obj"
+    } else if (crate_type == "rlib") {
+      output_file = "$target_out_dir/lib$crate_name.rlib"
+      emit_type = "link"
+    }
+    outputs += [ output_file ]
+    output_file_rel = rebase_path(output_file, root_build_dir)
+    args += [ "--emit=$emit_type=$output_file_rel" ]
+
+    # TODO(ry) For unknown reasons emitting a depfile on tests doesn't work.
```
Comment (ry) [inline build_extra/rust/rust.gni]: Remove

File: BUILD.gn (hunk 2 above)
Comment (piscisaureus) [reply BUILD.gn]: IDK - I thought of that too but I didn't try. I suspect its gonna be very brittle.

Another meh option is to generate a list of all crates in rust/BUILD.gn, and just pass that to all targets.

Maybe best to let it stew for a while and then ask some rust person for their advice.

File: build_extra/rust/get_rust_ldflags.py (hunk 3 above)
Comment (piscisaureus) [reply build_extra/rust/get_rust_ldflags.py]: Yeah I'll add a big comment to rust.gni and a small one here. You can't say this file is sparsely commented...

File: build_extra/rust/rust.gni (hunk 4 above)
Comment (piscisaureus) [reply build_extra/rust/rust.gni]: gn help exec_script

File: BUILD.gn (hunk 1 above)
Comment (piscisaureus) [reply BUILD.gn]: They are, but test_cc still needs it here, because it's not linked with rust_executable.

**Comment by piscisaureus** [comment]
They always were linked, you just didn't see it because rustc did it all
under the covers.

File: BUILD.gn (hunk 2 above)
Comment (ry) [reply BUILD.gn]: > best to let it stew for a while 
agree
> Another meh option is to generate a list of all crates in rust/BUILD.gn
I believe fuchsia does something like this.

File: BUILD.gn (hunk 1 above)
Comment (ry) [reply BUILD.gn]: Maybe add a comment to that effect

File: build_extra/rust/get_rust_ldflags.py (hunk 3 above)
Comment (piscisaureus) [reply build_extra/rust/get_rust_ldflags.py]: comment: done

File: BUILD.gn (hunk 1 above)
Comment (piscisaureus) [reply BUILD.gn]: done

File: build_extra/rust/rust.gni (hunk 5 above)
Comment (piscisaureus) [reply build_extra/rust/rust.gni]: done

**Pull request merged**